### PR TITLE
feature(unlock-app) : Generate qr

### PR DESCRIPTION
**Pull request by julien51**
# Description

The EthCC team needs a away to generate QR code for tickets themselves.
This PR adds an API call to generate QR code signed by us. 

I also added the ability to pass API keys as query strings for somplicity.
 

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: locksmith/src/utils/middlewares/auth.ts [hunk 1]
```diff
@@ -39,8 +39,35 @@ export function createAccessToken(user: User) {
   })
 }
 
+export const authenticateWithApiKey = async (req: any, token: string) => {
```
Comment (searchableguy) [inline locksmith/src/utils/middlewares/auth.ts:42]: You can type the request. :D

File: locksmith/src/routes/v2/ticket.ts [hunk 2]
```diff
@@ -39,4 +42,14 @@ router.post(
   }
 )
 
+router.get(
+  '/:network/:lockAddress/:tokenId/qr',
```
Comment (searchableguy) [inline locksmith/src/routes/v2/ticket.ts:46]: is `tokenId` synonymous to `keyId` here?

File: locksmith/src/routes/v2/ticket.ts [hunk 3]
```diff
@@ -39,4 +42,14 @@ router.post(
...
+  authenticatedMiddleware,
+  applicationOnlyMiddleware,
```
Comment (searchableguy) [inline locksmith/src/routes/v2/ticket.ts:48]: Is `applicationOnly` middleware intentional? This looks like endpoint which can be used by users as well.

File: unlock-app/src/components/interface/VerificationStatus.tsx [hunk 4]
```diff
@@ -87,6 +87,11 @@ export const VerificationStatus = ({ data, sig }: Props) => {
     return <InvalidKey reason="This ticket has expired" />
   }
 
+  // The current owner does not match the signed owner
+  if (!unlockKey) {
+    return <InvalidKey reason="The current owner does not match the signature!" />
+
```
Comment (searchableguy) [inline unlock-app/src/components/interface/VerificationStatus.tsx:93]: missing `}` here perhaps?

File: locksmith/src/controllers/v2/ticketsController.ts [hunk 5]
```diff
@@ -98,4 +106,36 @@ export class TicketsController {
...
+      ).replace('data:image/gif;base64,', '')
+      const img = Buffer.from(qrCode, 'base64')
+
+      response.writeHead(200, {
+        'Content-Type': 'image/gif',
+      })
+      return response.end(img)
+    } catch (err) {
+      logger.error(err)
+      return response.sendStatus(500)
```
Comment (searchableguy) [inline locksmith/src/controllers/v2/ticketsController.ts:138]: A generic message which details what failed would be good here. "Failed to generate QR code"

**Review by searchableguy — CHANGES_REQUESTED** [review]
LGTM but need more clarification on a few points above.

File: unlock-app/src/components/interface/VerificationStatus.tsx (hunk 4 above)
Comment (julien51) [reply unlock-app/src/components/interface/VerificationStatus.tsx:93]: Wow this should not be part of this PR. Removing it!

File: locksmith/src/routes/v2/ticket.ts (hunk 2 above)
Comment (julien51) [reply locksmith/src/routes/v2/ticket.ts:46]: Good point. I am changing it in other places too.

**Review by searchableguy — APPROVED** [review]

**Pull request merged**